Model/Indexer/AbstractIndexer.php

```diff
@@ -0,0 +1,198 @@
+use Magento\Framework\Mview\ActionInterface as MviewActionInterface;
+use Magento\Framework\Indexer\ActionInterface as IndexerActionInterface;
+use Nosto\NostoException;
+use Nosto\Tagging\Model\Indexer\Dimensions\AbstractDimensionModeConfiguration as DimensionModeConfiguration;
+use Nosto\Tagging\Helper\Account as NostoHelperAccount;
+use Nosto\Tagging\Helper\Scope as NostoHelperScope;
+use Nosto\Tagging\Logger\Logger as NostoLogger;
+use Magento\Store\Model\StoreDimensionProvider;
+use Magento\Indexer\Model\ProcessManager;
+use Nosto\Tagging\Util\Benchmark;
+use Magento\Framework\App\ObjectManager;
+use Nosto\Tagging\Model\Indexer\Dimensions\ModeSwitcherInterface;
+use Magento\Store\Model\Store;
+use Traversable;
+use ArrayIterator;
+use InvalidArgumentException;
+use UnexpectedValueException;
+
+abstract class AbstractIndexer implements DimensionalIndexerInterface, IndexerActionInterface, MviewActionInterface
+{
+    /** @var NostoHelperAccount */
+    private $nostoHelperAccount;
+
+    /** @var NostoHelperScope */
+    private $nostoHelperScope;
+
+    /** @var NostoLogger */
+    public $nostoLogger;
+
+    /** @var ProcessManager */
+    private $processManager;
+
+    /** @var DimensionProviderInterface */
+    private $dimensionProvider;
+
+    /**
+     * AbstractIndexer constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoLogger $nostoLogger
+     * @param DimensionProviderInterface $dimensionProvider
+     * @param ProcessManager|null $processManager
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope,
+        NostoLogger $nostoLogger,
+        DimensionProviderInterface $dimensionProvider,
+        ProcessManager $processManager = null
+    ) {
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoLogger = $nostoLogger;
+        $this->dimensionProvider = $dimensionProvider;
+        $this->processManager = $processManager;
+    }
+
+    /**
+     * Get ModeSwitcher class to later get the indexer mode
+     *
+     * @return ModeSwitcherInterface
+     */
+    abstract public function getModeSwitcher(): ModeSwitcherInterface;
+
+    /**
+     * Implement logic of single store indexing
+     *
+     * @param Store $store
+     * @param array $ids
+     */
+    abstract public function doIndex(Store $store, array $ids = []);
+
+    /**
+     * @return string
+     */
+    abstract public function getIndexerId(): string ;
+
+    /**
+     * @param array $ids
+     * @suppress PhanTypeMismatchArgument
+     */
+    public function doWork(array $ids = [])
+    {
+        $userFunctions = [];
+
+        switch ($this->getModeSwitcher()->getMode()) {
+            case DimensionModeConfiguration::DIMENSION_NONE:
+                $userFunctions[] = function () use ($ids) {
```

Comment:
I'm curious, why not this:

```
                /** @var DimensionProviderInterface $dimensionProvider */
                $dimensionProvider = ObjectManager::getInstance()->get(StoreDimensionProvider::class);
                /** @var Dimension[] $dimension  */
                foreach ($dimensionProvider->getIterator() as $dimension) {
                    /** @noinspection PhpUnhandledExceptionInspection */
                    $this->executeByDimensions($dimension, new ArrayIterator($ids));
                }
```

Reply:
There is no DimensionProvider in case of `none`.  So cannot follow same logic.